apic_ml2/neutron/plugins/ml2/drivers/cisco/apic/mechanism_apic.py

```diff
@@ -191,6 +192,44 @@ def name_wrapper(*args, **kwargs):
         return name_wrapper
 
 
+class KeystoneNotificationEndpoint(object):
+    filter_rule = oslo_messaging.NotificationFilter(
+        event_type='identity.project.update')
+
+    def __init__(self, mechanism_driver):
+        self._driver = mechanism_driver
+
+    def info(self, ctxt, publisher_id, event_type, payload, metadata):
+        LOG.debug("Keystone notification getting called!")
+
+        tenant_id = payload.get('resource_info')
+        # malformed notification?
+        if not tenant_id:
+            return None
+        # we only update tenants which has been created in APIC. For other
+        # case, its nameAlias will be set when the first network is being
+        # created under that tenant
+        if not self._driver.name_mapper.is_tenant_in_apic(tenant_id):
+            return None
+
+        new_tenant_name = self._driver.name_mapper.update_tenant_name(
+            tenant_id)
+        if new_tenant_name:
+            obj = {}
+            obj['tenant_id'] = tenant_id
+            apic_tenant_name = self._driver._get_tenant(obj)
+            apic_app_profile = self._driver._get_network_app_profile(obj)
+            if isinstance(apic_tenant_name, apic_mapper.ApicName):
```

Comment:
wouldn't it be clearer to just decide whether to update the Tenant object or the App Profile object based on the driver's "single_tenant_mode" flag?